src/org/zaproxy/zap/extension/ascanrulesBeta/InsecureHTTPMethod.java

```diff
@@ -1,455 +1,702 @@
+                } else {
+                    log.debug("Response does *not* match expected third party pattern");
+                }
+                is.close();
+                os.close();
+                bos.close();
+                socket.close();
+                return false;
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug(
+                            "Could not establish a socket connection to a third party using the CONNECT HTTP method: NULL socket returned, or non-200 response");
+                    log.debug("The status line returned: " + statusLine);
+                }
+                return false;
+            }
+        } catch (Exception e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Could not establish a socket connection to a third party using the CONNECT HTTP method", e);
+            }
+        }
+        return false;
+    }
+
+    private final MessageAndEvidence testHttpMethod(String httpMethod) {
+        //httpMethod = httpMethod.toLowerCase().trim();
+
+        //TODO:chose a body that exploits a vulnerability
+        String randomKey = RandomStringUtils.random(15, "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
+        String randomValue = RandomStringUtils.random(15, "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
+
+        String requestBody = '"'+randomKey+"\":\""+randomValue+'"'; //"randomKey":"randomValue"
+        if (httpMethod.equals(HttpRequestHeader.TRACK) || httpMethod.equals(HttpRequestHeader.TRACE) || httpMethod.equals(HttpRequestHeader.CONNECT)) {
+            /* We have separate methods for them */
+            return null;
+        }
+        final HttpRequestHeader traceRequestHeader = this.getBaseMsg().getRequestHeader();
+
+
+        traceRequestHeader.setMethod(httpMethod);
+        traceRequestHeader.setVersion(HttpRequestHeader.HTTP11);
+
+        final HttpMessage tracemsg = getNewMsg();
+        tracemsg.setRequestHeader(traceRequestHeader);
+        if(httpMethod.equals(HttpRequestHeader.PUT)||httpMethod.equals(HttpRequestHeader.PATCH)) {
+            tracemsg.setRequestBody(requestBody);
+        }
+        try {
+            // do not follow redirects. That might ruin our day.
+            sendAndReceive(tracemsg, false);
+        } catch (IOException e) {
+            log.error(e.getMessage(), e);
+            return null;
+        }
+
+        MessageAndEvidence ret = null;
+
+        final int responseCode = tracemsg.getResponseHeader().getStatusCode();
+
+        /*
+         * Build a list with status code which indicate that this HTTP method is
+         * enabled but we are not allowed to use it
+         */
+        final ArrayList<Integer> enabledStatusCodes = new ArrayList<>();
+        enabledStatusCodes.add(HttpStatusCode.UNAUTHORIZED);
+        enabledStatusCodes.add(HttpStatusCode.PAYMENT_REQUIRED);
+        enabledStatusCodes.add(HttpStatusCode.FORBIDDEN);

```

Comment:
405 - Method Not Allowed?
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405